File: WebDriverTests_Csharp/Autotest.Core/Autotests.Utilities/WebTestCore/TestSystem/ByExtensions/ByNthOfClass.cs
using System.Collections.ObjectModel;
using OpenQA.Selenium;

namespace Autotests.Utilities.WebTestCore.TestSystem.ByExtensions
{
    public class ByNthOfClass : By
    {
        private readonly string className;
        private readonly int index;

        internal ByNthOfClass(string className, int index)
        {
            this.className = className;
            this.index = index;
        }

        public override IWebElement FindElement(ISearchContext context)
        {
            ReadOnlyCollection<IWebElement> elements = ClassName(className).FindElements(context);
            if (index < 0 || index >= elements.Count)
                throw new NoSuchElementException(
                    string.Format("Element with class '{0}' and index '{1}' not found (actual count='{2}')", className,
                        index, elements.Count));
            return elements[index];
        }

        public override ReadOnlyCollection<IWebElement> FindElements(ISearchContext context)
        {
            ReadOnlyCollection<IWebElement> elements = ClassName(className).FindElements(context);
            if (index < 0 || index >= elements.Count)
                return new ReadOnlyCollection<IWebElement>(new IWebElement[0]);
            return new ReadOnlyCollection<IWebElement>(new[] {elements[index]});
        }

        public override string ToString()
        {
            return string.Format("ByNthOfClass: .{0}[{1}]", className, index);
        }
    }
}

File: WebDriverTests_Java/WebDriverTests/src/test/java/core/supportClass/ByNthOfClass.java
package core.supportClass;

import org.openqa.selenium.By;
import org.openqa.selenium.SearchContext;
import org.openqa.selenium.WebElement;

import java.util.Arrays;
import java.util.List;
import java.util.NoSuchElementException;
/**
 * Created by Виктория on 03.12.2015.
 */
public class ByNthOfClass extends By
{
    private String className;
    private int index;

    public ByNthOfClass(String className, int index)
    {
        this.className = className;
        this.index = index;
    }

    @Override
    public WebElement findElement(SearchContext context)
    {
        List<WebElement> elements = className(className).findElements(context);
        if (index < 0 || index >= elements.size())
        {
            throw new NoSuchElementException(String.format("Element with class '%1$s' and index '%2$s' not found (actual count='%3$s')", className, index, elements.size()));
        }
        return elements.get(index);
    }



    @Override
    public List<WebElement> findElements(SearchContext context)
    {
        List<WebElement> elements = className(className).findElements(context);
        if (index < 0 || index >= elements.size())
        {
            return Arrays.asList(new WebElement[0]);
        }
        WebElement[] e = new WebElement[] {elements.get(index)};
        return Arrays.asList( new WebElement[] {elements.get(index)});
    }


    @Override
    public String toString()
    {
        return String.format("ByNthOfClass: .%1$s[%2$s]", className, index);
    }
}
